components/ai_chat/core/common/proto_conversion.cc

```diff
@@ -0,0 +1,149 @@
+        !mojom_source->favicon_url.is_valid()) {
+      DLOG(ERROR) << "Invalid WebSourcesEvent found for persistence, with url: "
+                  << mojom_source->url.spec()
+                  << " and favicon url: " << mojom_source->favicon_url.spec();
+      continue;
+    }
+    store::WebSourceProto* proto_source = proto_event->add_sources();
+    proto_source->set_title(mojom_source->title);
+    proto_source->set_url(mojom_source->url.spec());
+    proto_source->set_favicon_url(mojom_source->favicon_url.spec());
+  }
+}
+
+mojom::ToolUseEventPtr DeserializeToolUseEvent(
+    const store::ToolUseEventProto& proto_event) {
+  auto mojom_event = mojom::ToolUseEvent::New(
+      proto_event.tool_name(), proto_event.id(), proto_event.arguments_json(),
+      std::nullopt, std::nullopt);
+
+  // Convert arguments from JSON string to mojo DictionaryValue
+  if (!proto_event.arguments_json().empty()) {
+    // We assume stored arguments is safe JSON and we don't need to parse
+    // in a separate process.
+    auto parsed_arguments =
+        base::JSONReader::ReadDict(proto_event.arguments_json());
+    if (parsed_arguments) {
+      mojom_event->arguments = std::move(parsed_arguments.value());
+    }
+  } else {
+    mojom_event->arguments = base::Value::Dict();
+  }
+
+  // Convert output ContentBlocks
+  if (proto_event.output_size() > 0) {
+    mojom_event->output = std::vector<mojom::ContentBlockPtr>();
+
+    for (const auto& proto_block : proto_event.output()) {
+      switch (proto_block.content_case()) {
+        case store::ContentBlockProto::kImageContentBlock: {
+          auto image_block = mojom::ImageContentBlock::New();
+          image_block->image_url =
+              GURL(proto_block.image_content_block().image_url());
+          mojom_event->output->push_back(
+              mojom::ContentBlock::NewImageContentBlock(
+                  std::move(image_block)));
+          break;
+        }
+        case store::ContentBlockProto::kTextContentBlock: {
+          auto text_block = mojom::TextContentBlock::New();
+          text_block->text = proto_block.text_content_block().text();
+          mojom_event->output->push_back(
+              mojom::ContentBlock::NewTextContentBlock(std::move(text_block)));
+          break;
+        }
+        case store::ContentBlockProto::CONTENT_NOT_SET:
+          // Skip invalid blocks
+          break;
+      }
+    }
+  }
+
+  return mojom_event;
+}
+
+bool SerializeToolUseEvent(const mojom::ToolUseEventPtr& mojom_event,
+                           store::ToolUseEventProto* proto_event) {
+  if (!mojom_event) {
+    return false;
+  }
+
+  proto_event->set_tool_name(mojom_event->tool_name);
+  proto_event->set_id(mojom_event->id);
+  proto_event->set_arguments_json(mojom_event->arguments_json);
+
+  // Convert output ContentBlocks
+  proto_event->clear_output();
+  if (mojom_event->output) {
+    for (const auto& mojom_block : *mojom_event->output) {
```

Comment:
why do we need to dereference `mojom_event` here?

Reply:
because `output` is optional
```
Invalid range expression of type 'std::optional<std::vector<ContentBlockPtr>>' (aka 'optional<vector<StructPtr<ContentBlock>>>'); did you mean to dereference it with '*'?
```